Support authenticating with AWS IAM role for MongoDB Atlas

gabrielcorado opened the pull request:
Closes #22972.

Supports [authenticating to MongoDB Atlas databases using AWS IAM roles](https://www.mongodb.com/docs/atlas/security/passwordless-authentication/), which consists of:
* Fetching access key, secret access key, and session token for the specified IAM Role.
* When connecting to MongoDB, use the Mongo AWS connector where:
    * Access key and secret access key are used as username and password.
    * Session token is passed as an additional property.

To connect, users can provide the ARN as a username:
```
$ tsh db connect --db-name=mongodb --db-user=arn:aws:iam::111111111111:role/mongodb-alice atlas
```

Or, a partial role ARN where Teleport will build the ARN using the account ID information fetched from database configuration or the agent credentials:
```
$ tsh db connect --db-name=mongodb --db-user=role/mongodb-alice atlas
```

<details>
<summary>Testing</summary>

1. AWS
    1. Create an IAM role for the database agent.
    2. Create an IAM Role to be used as the MongoDB user, and add the database agent at the trust policy with AssumeRole permission.
3. Atlas
    1. Create a database.
    2. Go to **"Database security"** and **"Add new database user"**. Select the **"AWS IAM"** and fill in the full ARN that will be used as the username and create the user.
5. Start the Teleport DB agent with the role credentials configured with the database configured:
    ```
    db_service:
      enabled: "yes"
      databases:
        - name: atlas
          protocol: "mongodb"
          uri: "mongodb+srv://teleport.xxxxxx.mongodb.net"
          tls: # Download the MongoDB Atlas CA: https://goteleport.com/docs/database-access/guides/mongodb-atlas/#atlas-ca-certificate
            ca_cert_file: "isrgrootx1.pem"
    ```

</details>

gabrielcorado committed:
feat: support authenticating with AWS IAM for MongoDB Atlas

gabrielcorado committed:
chore(lint): fix errors

gabrielcorado committed:
test(tsh): add missing database field

smallinsky commented on lib/srv/db/mongodb/connect.go:
```diff
@@ -153,6 +161,43 @@ func (e *Engine) getConnectionOptions(ctx context.Context, sessionCtx *common.Se
+	switch {
+	case sessionCtx.Database.GetType() == types.DatabaseTypeMongoAtlas && awsutils.IsRoleARN(sessionCtx.DatabaseUser):
+		e.Log.Debug("Authenticating to database using AWS IAM authentication")
+		username, password, sessToken, authErr := e.Auth.GetAtlasIAMToken(ctx, sessionCtx)
+		if err != nil {
+			return nil, trace.Wrap(authErr)
+		}
```
seems that error is assigned to `authErr` but the `if check`  is based on `err` variable: 
```suggestion
		username, password, sessToken, err := e.Auth.GetAtlasIAMToken(ctx, sessionCtx)
		if err != nil {
			return nil, trace.Wrap(err)
		}
```

smallinsky commented on lib/srv/db/mongodb/connect.go:
```diff
@@ -153,6 +161,43 @@ func (e *Engine) getConnectionOptions(ctx context.Context, sessionCtx *common.Se
+			Username: "CN=" + sessionCtx.DatabaseUser,
+		})
+	}
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
nit: 
I know that this helps for duplication but due to scope locality and readability i prefer sequence flow lie: function call  followed by error handling but leaving it up to you.

smallinsky commented on lib/srv/db/mongodb/connect.go:
```diff
@@ -153,6 +161,43 @@ func (e *Engine) getConnectionOptions(ctx context.Context, sessionCtx *common.Se
+	}
+
+	switch {
+	case sessionCtx.Database.GetType() == types.DatabaseTypeMongoAtlas && awsutils.IsRoleARN(sessionCtx.DatabaseUser):
```
the `sessionCtx.Database.GetType() == types.DatabaseTypeMongoAtlas` check is duplicated can we assign the result to local variable and reuse it ?

smallinsky commented on lib/srv/db/mongodb/connect.go:
```diff
@@ -153,6 +161,43 @@ func (e *Engine) getConnectionOptions(ctx context.Context, sessionCtx *common.Se
+
+	switch {
+	case sessionCtx.Database.GetType() == types.DatabaseTypeMongoAtlas && awsutils.IsRoleARN(sessionCtx.DatabaseUser):
+		e.Log.Debug("Authenticating to database using AWS IAM authentication")
```
```suggestion
		e.Log.Debug("Authenticating to database using AWS IAM authentication.")
```

smallinsky commented on lib/srv/db/common/auth.go:
```diff
@@ -822,6 +833,47 @@ func (a *dbAuth) getCurrentAzureVM(ctx context.Context) (*libazure.VirtualMachin
+		return "", "", "", trace.Wrap(err)
+	}
+
+	sess, err := a.cfg.Clients.GetAWSSession(ctx, dbAWS.Region, cloud.WithAssumeRole(arn, sessionCtx.Database.GetAWS().ExternalID))
```
What if the dbAWS.Region is empty ? 
I think that we support the case where a full ARN is provided and region is extracted from ARN.

gabrielcorado replied:
I've tested with an empty region, and it works fine. AFAIK the region is only used to determine the AWS partition (gov, china, or default); in that case, users can leave it empty if they are not on a different partition.

> I think that we support the case where a full ARN is provided and region is extracted from ARN.

The IAM role ARN doesn't contain the region, although it does have the partition. So I don't know if we should assume it from the agent ARN.

greedy52 replied:
Some UX questions:
1. should a mongo atlas database use fields from `aws` metadata block? I think this is fine. though not sure if UX is better if the mongo atlas section has its own `region`, `external_id` etc.
1. will mongo atlas support the aws cross-account feature?

greedy52 replied:
> I'm unsure if duplicating the configuration might be better since this AWS-specific integration could require something like aws_account_id and so on. So I'm not sure. However, we can improve the error messages and the documentation to mention which config users must fill.

👍 

> Yes, it does support (I've tested using a similar setup, except that the IAM role placed on the Atlas user was from an external account). However, I'm unsure how practical that would be since this integration requires no additional setup on the AWS account.

I don't see `dbAWS.AssumeRoleARN` being used here. Maybe I am missing something?

gabrielcorado replied:
> I don't see `dbAWS.AssumeRoleARN` being used here. Maybe I am missing something?

Is it required tho? In this case, we always assume the users' roles (from `tsh` command) as database usernames. However, it might require some change because it validates whether both `external_id` and `assume_role_arn` are provided.

I've tested with the following configuration, and it worked as expected. The teleport agent was using a role from account `111111111111`.
```yaml
- name: atlas
   protocol: "mongodb"
   uri: "mongodb+srv://tp.xxxxxxx.mongodb.net"
   tls:
     ca_cert_file: "isrgrootx1.pem"
   aws:
     account_id: "2222222222220"
     external_id: "example-external-id"
     assume_role_arn: "arn:aws:iam::222222222222:role/placeholder-role"
```

```code
$ tsh db connect --db-user=arn:aws:iam::222222222222:role/atlas-alice-role atlas
... success ...
```

gabrielcorado replied:
The agent assumes the `atlas-alice-role` directly. I've had to place the `assume_role_arn` in the configuration to fulfill the validation.

greedy52 replied:
When  `assume_role_arn` is to `placeholder-role`, it's expected that `placeholder-role` will assume `atlas-alice-role`. I would say it's more about role organization. Database agent assumes only the `placeholder-role` (most likely from a different AWS account) and `placeholder-role` has IAM permissions to other database access permissions like assuming `atlas-alice-role` in this case. This currently applies to all AWS databases.

I think we should either honour this logic for mongo atlas, or do not support `assume_role_arn` for atlas and return an error.

@gabrielcorado @smallinsky what do you think?

smallinsky replied:
@greedy52 
>  it's expected that placeholder-role will assume atlas-alice-role. 
Agree. we should support to role assumption chain. 

@gabrielcorado Could you provide a alignment for this ?

gabrielcorado replied:
I've updated it to do the role chaining.

smallinsky reviewed (commented):
First pass.

gabrielcorado committed:
refactor(mongodb): check for error on each authenticator branch

gabrielcorado committed:
refactor(mongodb): update log messages and atlas check

gabrielcorado committed:
refactor(auth): use IsRoleARN helper instead of IsARN

flyinghermit reviewed (approved):
Code looks good to me, but I haven't actually pulled this branch and tried to access MongoDB Atlas using AWS IAM roles. 

Please also get another approval from DB team.

gabrielcorado committed:
chore(db): remove unused line

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/mongo-atlas-iam

smallinsky reviewed (approved):
Nice, the integration looks really neat . 
I have left some nit comments also could you look  at CI flaky test detector  ? 
Otherwise LGTM.

greedy52 commented on lib/srv/db/common/test.go:
```diff
@@ -45,6 +45,8 @@ type TestServerConfig struct {
 	AuthUser string
 	// AuthToken is used in tests simulating IAM token authentication.
 	AuthToken string
+	// AuthSessionToken is used in tests simulating IAM token authentication.
+	AuthSessionToken string
```
nit: can this use the existing `AuthToken`?

greedy52 commented on lib/srv/db/mongodb/connect.go:
```diff
@@ -32,7 +32,18 @@ import (
+	awsSecretTokenKey = "AWS_SESSION_TOKEN"
+	// awsIAMSource is the authenticator source value used when authenticating
+	// using AWS IAM.
+	awsIAMSource = "$external"
```
nit: it would be great to provide a reference link, if there is one.

greedy52 reviewed (approved):
Tested and LGTM.

Few optional things that could be also helpful:
1. database user matcher
https://github.com/gravitational/teleport/blob/fe6e04461abd5a91960c35750dfe228e15596c43/lib/services/role.go#L2248-L2259
there were some extra logic that matches `db_users` in role definition vs `--db-user` flag so an AWS ARN can match a short role name. Probably can do something similar for this.

2. Since atlas endpoint is detected, could we download the CA pem automatically like other cloud databases?

And some UX questions https://github.com/gravitational/teleport/pull/26439/files#r1202941354


----
During test, I initially run into ` (AtlasError) bad auth : aws sts call has response 403`. I think it's a problem on Altas end as i recreated the database and everything went through.

gabrielcorado committed:
chore(mongodb): split authenticator func

gabrielcorado commented:
> 1. Database user matcher...

Do you mean using it to build the username or when checking access? If it's the second one, the MongoDB engine already uses this.

> Since atlas endpoint is detected, could we download the CA pem automatically like other cloud databases?

I've created an issue for this #26825.

gabrielcorado committed:
refactor(db): rename get atlas token function

gabrielcorado committed:
tests(db): reuse already existent auth property

gabrielcorado committed:
chore(mongodb): add docs reference

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/mongo-atlas-iam

gabrielcorado committed:
refactor(db): support role chaining

gabrielcorado committed:
feat(types): "require" iam role for atlas users

gabrielcorado commented:
> Seems atlas is a special case where database.RequireAWSIAMRolesAsUsers should be false since atlas can use "regular" user names, but it supports role/<name>

@greedy52 I agree it is a particular case, but the database username matcher for this case would also work as it matches the "direct" username. Should we do some other function or rename this to be more accurate?

greedy52 commented on lib/srv/db/common/auth.go:
```diff
@@ -838,29 +838,43 @@ func (a *dbAuth) GetAWSIAMCreds(ctx context.Context, sessionCtx *Session) (strin
+		return "", "", "", trace.Wrap(err)
+	}
+
+	sess, err := a.cfg.Clients.GetAWSSession(ctx, dbAWS.Region, cloud.WithChainedAssumeRole(baseSession, arn, dbAWS.ExternalID))
```
Thanks for updating this!

One nit comment. So when both `assume_role_arn` and `external_id` are specified, `external_id` is only used for the `assume_role_arn` and the second/chained assume role should use empty for external id. 

We had a debate on this as this behaviour is kinda confusing. But having two external ids can be confusing too. The current logic basically assumes `external_id` should be used for the first assume-role, and the second assume-role shouldn't need the external id as they should be in the same account.

https://github.com/gravitational/teleport/blob/f4385c16df26c4e140371cf15f770ff7bfd75158/lib/srv/db/dynamodb/engine.go#L214-L216

gabrielcorado replied:
Done. I've added some tests to ensure the behavior too.

gabrielcorado committed:
refactor(db): use external id only on the first session

gabrielcorado committed:
refactor(services): add new database matcher for regular users and aws

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/mongo-atlas-iam

gabrielcorado commented:
> > @greedy52 I agree it is a particular case, but the database username matcher for this case would also work as it matches the "direct" username. Should we do some other function or rename this to be more accurate?
> 
> Setting `RequireAWSIAMRolesAsUsers` to true for mongodb atlas does work for role ARN and `role/<name>` matching, but it also raises a security concern on non-AWS roles. For example `tsh db connect --db-user alice atlas` (expecting to connect x509 `CN=alice` would get allowed by `arn:aws:iam::123456789012:role/alice`.

Got it. I've updated it to have a separate flow (called "support aws roles"). If you could look, it is in [this commit](https://github.com/gravitational/teleport/pull/26439/commits/a24cdb66ed3da75da023f0a26031c5b3026995be). @smallinsky If you could take a look, too, that would be great. Thanks.

greedy52 commented on lib/services/role.go:
```diff
@@ -2332,6 +2332,13 @@ type databaseUserMatcher struct {
 // NewDatabaseUserMatcher creates a RoleMatcher that checks whether the role's
 // database users match the specified condition.
 func NewDatabaseUserMatcher(db types.Database, user string) RoleMatcher {
+	if db.SupportAWSIAMRolesAsUsers() {
+		return &databaseUserMatcher{
+			user:             user,
+			alternativeNames: makeUsernamesForAWSRole(db, user),
+		}
+	}
```
nit: the logic looks good to me. maybe we can improve the naming to differentiate between `makeAlternativeNamesForAWSRole`, something like
```suggestion
	if db.SupportAWSIAMRoleARNsAsUsers() {
		return &databaseUserMatcher{
			user:             user,
			alternativeNames: makeUsernamesForAWSRoleARN(db, user),
		}
	}
```

gabrielcorado committed:
chore(db): rename functions to be more assertive

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/mongo-atlas-iam

gabrielcorado committed:
chore(types): fix lint

gabrielcorado committed:
test(db): remove duplicated test

The test being removed here is covered by `TestMongoDBAtlas`
(lib/srv/db/auth_test.go).
